model/src/guild/auto_moderation/rule.rs

```diff
@@ -0,0 +1,40 @@
+use super::{
+    AutoModerationAction, AutoModerationEventType, AutoModerationTriggerMetadata,
+    AutoModerationTriggerType,
+};
+use crate::id::{
+    marker::{AutoModerationRuleMarker, ChannelMarker, GuildMarker, RoleMarker, UserMarker},
+    Id,
+};
+use serde::{Deserialize, Serialize};
+
+/// Configured auto moderation rule.
+#[derive(Clone, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
+pub struct AutoModerationRule {
+    /// Actions which will execute when the rule is triggered.
+    pub actions: Vec<AutoModerationAction>,
+    /// User which first created the rule.
+    pub creator_id: Id<UserMarker>,
+    /// Whether the rule is enabled.
+    pub enabled: bool,
+    /// Rule event type.
+    pub event_type: AutoModerationEventType,
+    /// Channel ids that should not be affected by the rule.
+    ///
+    /// Maximum of 50.
+    pub exempt_channels: Vec<Id<ChannelMarker>>,
+    /// Role IDs that should not be affected by the rule.
+    ///
+    /// Maximum of 20.
+    pub exempt_roles: Vec<Id<RoleMarker>>,
+    /// ID of the guild the rule belongs to.
+    pub guild_id: Id<GuildMarker>,
+    /// ID of the rule.
+    pub id: Id<AutoModerationRuleMarker>,
+    /// Name of the rule.
+    pub name: String,
+    /// Rule trigger metadata.
+    pub trigger_metadata: AutoModerationTriggerMetadata,
+    /// Rule trigger type.
+    pub trigger_type: AutoModerationTriggerType,
```

Comment:
Merge into an enum:
```rs
pub enum AutoModerationTrigger {
    Keyword(Vec<String>),
    HarmfulLink,
    Spam,
    KeywordPreset(Vec<AutoModerationKeywordPresetType>),
    Unknown(u8),
}
```

Reply:
i'll have to think more about this one, it doesn't follow the discord model and requires custom deser, which we are trying to avoid

Reply:
It does not, but not merging tags and values, essentially never using non-unit enums, is a drastic change and should therefore be subject to an RFC so that the impact is better understood.
Enums are from my point of view a necessary compromise between usability, and internal code complexity and API future proofing. Twilight is first and foremost a Rust library, so users expect Rust idioms over strict Discord API adherence.

Reply:
From a user point of view, the point of twilight is to be low-level and to have fewer abstractions (kind of?) as possible. I would think that merging two fields "abstracts" away the original object, which if so, no offense, might be exactly opposite to what twilight aims to be.

Reply:
Discord sends tagged unions (union + tag), which is exactly what Rust enums are. We too could model it using unions directly https://doc.rust-lang.org/std/keyword.union.html

Reply:
problem is, serde can only automatically serialize the type name, not as a number. I think we would need to do ugly `#[serde(rename = "0")]` which i am not sure about